### PR TITLE
🚀 Cached HTML renderer

### PR DESCRIPTION
**Pull request by alanorozco**
Since all templates are static, we can key a rendered tree cache by the template string itself. [Cloning nodes from said cache is significantly faster than setting `innerHTML` each time,](http://jsbench.github.io/#c7cf6f598b80392bf4f42c932f703fcd) even when using our very own `LruCache` over a simple dict. Ops/second, so higher is better:

image

- Introduces `cachedHtmlFor` for explicit use of cached templates in cases where render is likely to occur more than once during a short sequence of frames.

- Introduces `TempCache` for a time-based eviction cache.

- Updates `src` modules to use `cachedHtmlFor` where it makes sense.

- Refactors `static-template` into a service in order to maintain cache (and rendering container!) across extensions.

- Updates lint rules and transforms.

## Drawbacks

This is memory-heavy and should be used sparingly. See comments on `cachedHtmlFor` for details.

## Known issues/Open questions

- I'm not sure if there are any lint rules tests that I need to update or test cases that I need to check manually. (@jridgewell)
- A few extension tests failing per no service bootstrapping/doc scopes which should be easily fixed.

### CODE REVIEW

**Comment by alanorozco** [comment]
@jridgewell Right, that's the tradeoff. Rationale is that by making cache usage explicit, we're limiting the surface of keys and explicitly selecting cases were the node would take less memory (i.e. shallow trees).

I agree that `LruCache` is not the best structure for this but disagree on the approach to loops. The majority of cache hits with this method would be async (loaders, sizers, specific cases that respond to component events, etc.) so limiting this to loops wouldn't make sense.

**Comment by alanorozco** [comment]
@jridgewell How about a wrapper around `LruCache` that flushes after a certain condition (could be time, could be macro tasks)?

**Comment by jridgewell** [comment]
> and explicitly selecting cases were the node would take less memory (i.e. shallow trees).

DOM nodes are inherently memory intensive. Even a simple `"<div></div>"` will take orders of more memory than an plain JS object..

> The majority of cache hits with this method would be async (loaders, sizers, specific cases that respond to component events, etc.) so limiting this to loops wouldn't make sense.

These are all time latent then. Does it really matter if an operation takes 9.446μs (0.000010092s vs 0.000000646s, from the jsbench page) longer if it's only run every 16ms? It's a savings of just 0.0590375%. If the loaders are being created even less frequently, than we're not really saving any real time, but we are keeping memory alive.

**Comment by alanorozco** [comment]
@jridgewell 

> These are all time latent then. Does it really matter if an operation takes 9.446μs (0.000010092s vs 0.000000646s, from the jsbench page) longer if it's only run every 16ms?

Maybe I'm misunderstanding something, but there are cases where many loaders are rendered during the same frame (e.g. render a bunch of slow images that all fit in viewport).

**Comment by jridgewell** [comment]
Those are templates happening in a loop (a really deep one through `Resources`), and definitely something we should think about optimizing! With a `TempCache` that clears after a few macro tasks, those renders would hit the cached path.

But, we still have to think whether that's worth it, too. It'd take 20 loader renders in a single frame just to have a 1% impact on the 16ms frame budget. I'm not sure that's worth any additional complexity.

**Comment by alanorozco** [comment]
A 1%  _time_ impact on our very fast machines running `jsbench` 😉

Loaders are just a single case, but there are many different templates that could be rendered and cached during the same frame.

We could run a task-based eviction cache under an experiment and see if that makes any difference?

**Comment by jridgewell** [comment]
A 6x CPU slowdown gives 77μs vs 6μs, for a savings of 71μs and 0.44% of our 16ms budget. That makes it more palatable. I'll wait for everyone else's opinion.

**Comment by jridgewell** [comment]
> There is a small CPU perf win with a small memory usage increase. We are CPU bound and memory footprint of AMP is tiny compared to an average web page, so we are way below limits of even worrying about it.

The issue here is `LRUCache`, the fact that it is a permanent (until at capacity) cache, and the _size_ of the memory increase. Take the loading dots template:

```js
html`<div class="i-amphtml-loader">
  <div class="i-amphtml-loader-dot"></div>
  <div class="i-amphtml-loader-dot"></div>
  <div class="i-amphtml-loader-dot"></div>
</div>`
```

[Chrome reports](https://gist.github.com/jridgewell/b49c21b13da2863179fee9d1c080a577) a **31784 byte** memory increase for this _simple_ template. Caching DOM isn't cheap. We're saving microseconds over the lifetime of a page at the expense of KB of memory.

To move forward, this needs to implement a `TempCache` that's bound by a maximum time, not capacity. Then we can debate whether microseconds are worth the complexity.

**Comment by kristoferbaxter** [comment]
This is an interesting idea, and I want to think about the caching strategy a bit more before commenting further.

I agree this would be valuable in tight loops.

**Comment by alanorozco** [comment]
Went ahead and implemented a `TempCache`. Looking forward to additional feedback about caching strategies or this approach in general.

[Here's a demo of the cached renderer using `TempCache`](https://html-temp-cache.glitch.me/). See console for cache hit/flush logs. Scroll doc for some interesting results. The amount of macro-tasks to wait can be set as a fragment, e.g. `#flush=5` (default 3).

File: src/static-template.js [hunk 1]
```diff
@@ -14,52 +14,132 @@
...
+      cache.put(key, htmlInternal(strings));
+    return seed.cloneNode(/* deep */ true);
+  };
+
+  /**
+   * @param {!Array<string>} strings
+   * @return {!Element}
+   */
+  const htmlInternal = strings => {
+    devAssert(strings.length === 1, 'Improper html template tag usage.');
```
Comment (jridgewell) [inline src/static-template.js]: Duplicate the dev assert in `cachedHtmlInternal`

File: src/utils/temp-cache.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    /** @private {?Object<(number|string), T>} */
+    this.cache_ = null;
+  }
+
+  /**
+   * @return {boolean}
+   * @private
+   */
+  time_() {
+    return (new this.win_.Date()).getTime();
```
Comment (jridgewell) [inline src/utils/temp-cache.js]: `Date.now()`

File: src/utils/temp-cache.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    this.flushAfterMacroTasks_ = flushAfterMacroTasks;
+
+    /** @private @const {function()} */
+    this.flush_ = () => {
+      dev().info(TAG, `flush [${this.time_()}]`);
+      this.cache_ = null;
+    };
+
+    /** @private {?Object<(number|string), T>} */
+    this.cache_ = null;
```
Comment (jridgewell) [inline src/utils/temp-cache.js]: Just initialize the cache to an object.

File: src/utils/temp-cache.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    /** @private @const {!Window} */
+    this.win_ = win;
+
+    /** @private @const {number} */
+    this.flushAfterMacroTasks_ = flushAfterMacroTasks;
+
+    /** @private @const {function()} */
+    this.flush_ = () => {
+      dev().info(TAG, `flush [${this.time_()}]`);
+      this.cache_ = null;
```
Comment (jridgewell) [inline src/utils/temp-cache.js]: Flush should just recreate the object.

File: src/utils/temp-cache.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  time_() {
+    return (new this.win_.Date()).getTime();
+  }
+
+  /**
+   * @param {number|string} key
+   * @return {boolean}
+   */
+  has(key) {
+    return this.cache_ && this.cache_[key] !== undefined;
```
Comment (jridgewell) [inline src/utils/temp-cache.js]: No need for cache check, once it's already initialized.

File: src/utils/temp-cache.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  /**
+   * @param {number|string} key
+   * @return {T|undefined}
+   */
+  get(key) {
+    if (!this.has(key)) {
```
Comment (jridgewell) [inline src/utils/temp-cache.js]: No need for has check once the object is already initialized.

File: src/utils/temp-cache.js (hunk 5 above)
Comment (jridgewell) [inline src/utils/temp-cache.js]: `!!this.cache_[key]`, same as `LRUCache`

File: src/utils/temp-cache.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  /**
+   * @return {!Object<(number|string), T>}
+   * @private
+   */
+  maybeInitialize_() {
+    const cache = this.cache_;
+    if (cache) {
+      return cache;
+    }
+    waitForMacroTasks(this.win_, this.flush_, this.flushAfterMacroTasks_);
```
Comment (jridgewell) [inline src/utils/temp-cache.js]: Just use a regular `setTimeout` for X ms.

File: test/unit/test-static-template.js [hunk 8]
```diff
@@ -47,7 +54,8 @@ describe('Static Template', () => {
       expect(p.nextSibling).to.be.null;
     });
 
-    it('creates tree with last ownerDocument', () => {
+    // TODO(alanorozco): Unsure about what to do with this test.
```
Comment (jridgewell) [inline test/unit/test-static-template.js]: Delete it now that it's a service.

**Comment by choumx** [comment]
What may help ground the discussion is a quick test on current usages on real pages.

Looking through existing usages of `htmlFor()`, seems like it could speed up loader element/container creation in loader.js/custom-element.js.

In extensions, `amp-story` seems like the most promising beneficiary here. `amp-lightbox` has a loop in `buildCarouselSlides_()`. Remaining usages seem like one-and-done's.

File: src/static-template.js [hunk 9]
```diff
@@ -14,52 +14,132 @@
...
+
+/**
+ * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrDoc
+ * @return {{
+ *  htmlInternal: !HtmlLiteralTagDef,
+ *  cachedHtmlInternal: !HtmlLiteralTagDef,
+ * }}
+ */
+function getHtml(nodeOrDoc) {
+  return getServiceForDoc(getAmpdoc(nodeOrDoc), SERVICE);
```
Comment (jridgewell) [inline src/static-template.js]: Wait, didn't we prevent passing an `ampdoc` into a `getServiceForDoc` call? @choumx

File: src/static-template.js [hunk 10]
```diff
@@ -14,52 +14,132 @@
...
+ * @param {!./service/ampdoc-impl.AmpDoc} ampdoc
+ * @return {{
+ *  htmlInternal: !HtmlLiteralTagDef,
+ *  cachedHtmlInternal: !HtmlLiteralTagDef,
+ * }}
  */
-function html(strings) {
-  devAssert(strings.length === 1, 'Improper html template tag usage.');
-  container./*OK*/innerHTML = strings[0];
+export function buildHtmlService(ampdoc) {
```
Comment (jridgewell) [inline src/static-template.js]: This can just be made into a class?

File: src/static-template.js (hunk 9 above)
Comment (alanorozco) [reply src/static-template.js]: Looking forward to knowing what I did wrong and fix it 🤓

Comment (choumx) [reply src/static-template.js]: It's incomplete. Most of the getters in services.js are migrated, but not all so we can't change `getServiceForDoc` yet.

But yea, change `nodeOrDoc` to `element`.

File: src/static-template.js [hunk 11]
```diff
@@ -95,51 +93,58 @@ export function cachedHtmlFor(nodeOrDoc) {
...
+    this.container_ = ampdoc.getRootNode().createElement('div');
+
+    /** @type {?TempCache<!Element>} */
+    this.cache_ = null;
+
+    /**
+     * @param {!Array<string>} strings
+     * @return {!Element}
+     */
+    this.cachedHtmlInternal = strings => { // must be arrow func for scope
```
Comment (jridgewell) [inline src/static-template.js]: These can be regular methods on the class.

File: src/utils/temp-cache.js [hunk 12]
```diff
@@ -79,19 +70,15 @@ export class TempCache {
    * @return {T} Returns payload back
    */
   put(key, payload) {
-    return (this.maybeInitialize_()[key] = payload);
+    this.maybeSetFlushTimeout_();
+    return (this.cache_[key] = payload);
```
Comment (jridgewell) [inline src/utils/temp-cache.js]: Nit: break this into two lines. The minifier will do the magic.

File: src/static-template.js (hunk 11 above)
Comment (alanorozco) [reply src/static-template.js]: Tried i. By passing e.g. `getServiceForDoc(element, SERVICE).htmlInternal` method becomes unbound and breaks.

File: src/utils/temp-cache.js (hunk 12 above)
Comment (alanorozco) [reply src/utils/temp-cache.js]: Nice.

**Comment by alanorozco** [comment]
@jridgewell I don't have any interesting cases at hand, but feel free to run this locally and proxy some cached docs.

Alternatively, [I just put up a demo](https://html-cache-infinite-scroll.glitch.me/) that's a combination of the infinite scroll demo shown internally on Tuesday and this caching method. You can also set e.g. `#flush=200` to set the amount of time in `ms` to wait before flush.

File: src/utils/lru-cache.js [hunk 13]
```diff
@@ -69,13 +69,15 @@ export class LruCache {
...
    * @param {T} payload The payload to cache.
+   * @return {T} Returns payload back
    */
   put(key, payload) {
     if (!this.has(key)) {
       this.size_++;
     }
     this.cache_[key] = {payload, access: this.access_};
     this.evict_();
+    return payload;
```
Comment (kristoferbaxter) [inline src/utils/lru-cache.js]: Intentional change since this PR is no longer using the LRU cache?

Comment (alanorozco) [reply src/utils/lru-cache.js]: Yeah, kept it for convenience and consistency.

**Comment by alanorozco** [comment]
@choumx Kept cached usages where there's explicit loops and loaders. PTAL.

**Comment by mrjoro** [comment]
Looks like this is pending review from @choumx @kristoferbaxter and @aghassemi?

**Comment by alanorozco** [comment]
This is still active, blocked by me. Haven't had time to get back to it.

**Pull request closed without merging**